src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -45,40 +53,65 @@ public VisualStudioProjectFactory(
         [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> fileInfoProviders,
         IVisualStudioDiagnosticAnalyzerProviderFactory vsixAnalyzerProviderFactory,
         [ImportMany] IEnumerable<IAnalyzerAssemblyRedirector> analyzerAssemblyRedirectors,
-        IVsService<SVsSolution, IVsSolution2> solution2)
+        IVsService<SVsSolution, IVsSolution2> solution2,
+        IAsynchronousOperationListenerProvider listenerProvider)
     {
         _threadingContext = threadingContext;
         _visualStudioWorkspaceImpl = visualStudioWorkspaceImpl;
         _dynamicFileInfoProviders = fileInfoProviders.AsImmutableOrEmpty();
         _vsixAnalyzerProviderFactory = vsixAnalyzerProviderFactory;
         _analyzerAssemblyRedirectors = analyzerAssemblyRedirectors.AsImmutableOrEmpty();
         _solution2 = solution2;
-    }
 
-    public Task<ProjectSystemProject> CreateAndAddToWorkspaceAsync(string projectSystemName, string language, CancellationToken cancellationToken)
-        => CreateAndAddToWorkspaceAsync(projectSystemName, language, new VisualStudioProjectCreationInfo(), cancellationToken);
+        _uiContextUpdateWorkQueue = new AsyncBatchingWorkQueue<string>(
+            DelayTimeSpan.NearImmediate,
+            UpdateUIContextsAsync,
+            StringComparer.OrdinalIgnoreCase,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            _threadingContext.DisposalToken);
+
+        _initializationTask = _threadingContext.JoinableTaskFactory.RunAsync(
+            async () =>
+            {
+                var cancellationToken = _threadingContext.DisposalToken;
+                // HACK: Fetch this service to ensure it's still created on the UI thread; once this is
+                // moved off we'll need to fix up it's constructor to be free-threaded.
+
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
```

Comment:
any specific reason for alwaysYield?

Reply:
I didn't know about the costs of the code following it (specifically getting the VisualStudioMetadataReferenceManager) and didn't want to block construction of VisualStudioProjectFactory if it were called on the main thread.

Reply:
Done!